### PR TITLE
update: add open_by_handle_at syscall

### PR DESCRIPTION
**Pull request by Andreagit97**
**What type of PR is this?**

/kind feature

**Any specific area of the project related to this PR?**

/area driver-kmod

/area driver-ebpf

/area libscap

/area libsinsp

**What this PR does / why we need it**:

This adds support to the `open_by_handle_at` syscall. This has been reported as a high priority missing syscall in the "umbrella issue" :point_down:  

https://github.com/falcosecurity/falco/issues/676

**Which issue(s) this PR fixes**:

**Special notes for your reviewer**:

The following sample program has been used for testing:

[main.c.txt](https://github.com/falcosecurity/libs/files/7648568/main.c.txt)

1. Compile `main.c`:

```
gcc main.c -o main
```

2. Create the file to open where you want.

3. Run `main` passing as argument the absolute file path or the path relative to the current working directory of the calling process:

```
./main <path>
```

**Open_by_hanlde_at path extraction**

We have chosen to extract the absolute path of the opened file from the file descriptor returned by the syscall. In case of error, the file descriptor is not valid so unfortunately, we cannot extract the file path. The file descriptor seems the smartest way to extract information about the path, given the little information provided by the syscall.

**Bpf testing**

I have tested the new bpf features on the following vagrant images:

1. Image = bento/amazonlinux-2
- Kernel = 4.14.198-152.320.amzn2.x86_64
- clang = clang version 11.1.0 

2. Image = generic/debian10
- Kernel = 4.19.0-18-amd64
- clang = clang version 7.0.1-8+deb10u2 

3. Image = generic/centos8
- Kernel = 4.18.0-348.2.1.el8_5.x86_64
- clang = clang version 12.0.1 

4. Image = generic/fedora33
- Kernel = 5.14.18-100.fc33.x86_64
- clang = clang version 11.0.0 (Fedora 11.0.0-3.fc33)

5. Image = generic/alpine314
- Kernel = 5.10.82
- clang = clang version 11.1.0

6. Image = bento/ubuntu-20.10
- Kernel = 5.8.0-63-generic
- clang = clang version 11.0.0-2

7. Image = bento/ubuntu-21.04
- Kernel = 5.14.0-051400-generic
- clang = clang version 12.0.0

8. Image = opensuse/Tumbleweed.x86_64
- Kernel = 5.15.6
- clang = clang version 13.0.0

Old kernels verifiers (like 4.14) allow reading only file paths consisting of at most 16 components. With more than 16 components the verifier complains about the excessive bpf program length due to the loop unrolling approach used in the new `bpf_get_path` function. The function `bpf_get_path` added in this pull request, is an attempt to collect the file path from the file descriptor. Kernel 5.10 introduced a new bpf_helper called `bpf_d_path` which however cannot be used by `BPF_PROG_TYPE_RAW_TRACEPOINT` programs. Unfortunately, in our case, libscap loads all bpf programs as `BPF_PROG_TYPE_RAW_TRACEPOINT`.

**Does this PR introduce a user-facing change?**:

```release-note
update: add support for `open_by_handle_at` syscall
```

### CODE REVIEW

File: driver/bpf/filler_helpers.h [hunk 1]
```diff
@@ -56,6 +59,57 @@ static __always_inline struct file *bpf_fget(int fd)
 	return fil;
 }
 
+// Kernel 5.10 introduced a new bpf_helper called `bpf_d_path` to extract a file path starting from a file descriptor.
+// Libscap loads our bpf programs as `BPF_PROG_TYPE_RAW_TRACEPOINT` programs. This type of program doesn't seem able to call this new helper because it is out of its scope. For more details see here https://github.com/torvalds/linux/blob/58e1100fdc5990b0cc0d4beaf2562a92e621ac7d/kernel/trace/bpf_trace.c#L1574
+static __always_inline char *bpf_get_path(struct filler_data *data, int fd)
```
Comment (jasondellaluce) [inline driver/bpf/filler_helpers.h:64]: This function looks good, but we might want to test it across multiple clang/bpf-verifier combinations to make sure that it does not break the probe on certain kernel versions.

File: driver/bpf/fillers.h [hunk 2]
```diff
@@ -2389,6 +2389,67 @@ FILLER(sys_openat2_x, true)
...
+		val = PPM_AT_FDCWD;
+	}
+	
+	res = bpf_val_to_ring(data, val);
+	if (res != PPM_SUCCESS)
+	{
+		return res;
+	}
+
+    /*
```
Comment (jasondellaluce) [inline driver/bpf/fillers.h:2424]: Indentation seems broken here and at line 2427.

File: userspace/libsinsp/filterchecks.cpp [hunk 3]
```diff
@@ -3312,6 +3328,29 @@ uint8_t *sinsp_filter_check_event::extract_abspath(sinsp_evt *evt, OUT uint32_t
...
+		ASSERT(parinfo->m_len == sizeof(int64_t));
+		fd = *(int64_t *)parinfo->m_val;
+
+		if(fd>0)
+		{
+			//
+			// Get the file path directly from the ring buffer.
+			//
+			parinfo = evt->get_param(3);
+	        strcpy(fullname, parinfo->m_val);
```
Comment (jasondellaluce) [inline userspace/libsinsp/filterchecks.cpp:3349]: Indentation seems broken here and in line 3351.

File: driver/bpf/filler_helpers.h (hunk 1 above)
Comment (Andreagit97) [reply driver/bpf/filler_helpers.h:65]: I have tested the bpf function for multiple clang/bpf-verifier combinations, see the PR description for more details.

**Review by leogr — APPROVED** [review]
/approve

**Pull request merged**